pageserver/src/tenant/timeline/detach_ancestor.rs

```diff
@@ -116,6 +130,472 @@ impl Default for Options {
+            return Err(Error::ShuttingDown);
+        }
+
+        let mut guard = self.inner.lock().unwrap();
+        let completion = guard.start_new(&detached.timeline_id)?;
+        // now that we changed the contents, notify any long-sleeping gc
+        self.gc_waiting.notify_one();
+        Ok(completion)
+    }
+
+    async fn wait_until_gc_is_paused(&self, detached: &Arc<Timeline>) -> Result<(), Error> {
+        let mut cancelled = std::pin::pin!(detached.cancel.cancelled());
+
+        loop {
+            tokio::select! {
+                _ = &mut cancelled => { return Err(Error::ShuttingDown); },
+                _ = self.attempt_waiting.notified() => {},
+            };
+
+            // reading a notification which was not intended for us is not a problem,
+            // because we check if *our* progress has been witnessed by gc.
+            let g = self.inner.lock().unwrap();
+            if g.is_gc_paused(&detached.timeline_id) {
+                return Ok(());
+            }
+        }
+    }
+
+    /// Completes a previously started detach ancestor attempt. To be called *after* the operation
+    /// including the tenant has been restarted. The completion is persistent, and no reparentings
+    /// can be done afterwards.
+    ///
+    /// Cancellation safe.
+    pub(crate) async fn complete(
+        &self,
+        attempt: Attempt,
+        tenant: &Arc<Tenant>,
+    ) -> Result<(), Error> {
+        // do we need the tenant to actually activate...? yes we do, but that can happen via
+        // background task starting, because we will similarly want to confirm that the gc has
+        // paused, before we unpause it?
+
+        // assert that such and such state has been collected
+        // find the timeline the attempt represents
+        // using the timelines remote client, upload an index part with completion information
+
+        self.inner.lock().unwrap().validate(&attempt);
+
+        // FIXME: what if this fails, what do we do then?
+        tenant
+            .wait_to_become_active(std::time::Duration::from_secs(9999))
+            .await
+            .map_err(|e| {
+                use pageserver_api::models::TenantState;
+                use GetActiveTenantError::{Cancelled, WillNotBecomeActive};
+                match e {
+                    Cancelled | WillNotBecomeActive(TenantState::Stopping { .. }) => {
+                        Error::ShuttingDown
+                    }
+                    other => Error::Complete(other.into()),
+                }
+            })?;
+
+        utils::pausable_failpoint!(
+            "timeline-detach-ancestor::after_activating_before_finding-pausable"
+        );
+
+        // no need to cancel the attempt, because timeline deletion and/or tenant restart has
+        // already marked it unblocked. TODO: would be nice to cancel just in case though.
+        let timeline = tenant
+            .get_timeline(attempt.timeline_id, true)
```

Comment:
This once could fail if the timeline gets deleted inbetween. Current cplane won't do it but highlights a problem. We could avoid this case by continuing to carry the slot_guard from the caller.